src/ast/ast_validations.c

```diff
@@ -1113,6 +1118,92 @@ static VISITOR_STRATEGY _Validate_CALL_Clause
 	return VISITOR_CONTINUE;
 }
 
+static void _AST_GetWithIdentifiers(
+	const cypher_astnode_t *node, 
+	rax *identifiers
+) {
+	if(!node) return;
+	ASSERT(identifiers != NULL);
+
+	cypher_astnode_type_t type = cypher_astnode_type(node);
+	if(type == CYPHER_AST_IDENTIFIER) {
+		const char *identifier = cypher_ast_identifier_get_name(node);
+		raxInsert(identifiers, (unsigned char *)identifier, strlen(identifier), NULL, NULL);
+		return;
+	}
+
+	// Don't visit the children of pattern comprehensions, as these will be aliased later.
+	if(type == CYPHER_AST_PATTERN_COMPREHENSION) return;
+
+	uint child_count = cypher_astnode_nchildren(node);
+
+	// In case current node is of type projection inspect first child only
+	if(type == CYPHER_AST_PROJECTION) child_count = 1;
+
+	for(uint i = 0; i < child_count; i++) {
+		const cypher_astnode_t *child = cypher_astnode_get_child(node, i);
+		_AST_GetWithIdentifiers(child, identifiers);
+	}
+}
+
+static void _AST_GetWithReferences
+(
+	const cypher_astnode_t *node, 
+	rax *identifiers
+) {
+	if(!node) return;
+	if(cypher_astnode_type(node) != CYPHER_AST_WITH) return;
+	ASSERT(identifiers != NULL);
+
+	uint num_with_projections = cypher_ast_with_nprojections(node);
+	for(uint i = 0; i < num_with_projections; i ++) {
+		const cypher_astnode_t *child = cypher_ast_with_get_projection(node, i);
+		_AST_GetWithIdentifiers(child, identifiers);
+	}
+}
+
+static AST_Validation _Validate_Aliases_DefinedInWithClause(
+	const cypher_astnode_t *clause,
+	rax *defined_aliases
+) {
+	AST_Validation res = AST_VALID;
+	rax *with_aliases = raxNew();
+	rax *with_referred = raxNew();
+
+	raxIterator it;
+
+	// Get aliases identifiers in with clause
+	_AST_GetWithAliases(clause, with_aliases);
+
+	// Get referred identifiers in with clause, but don't include references of ORDER BY
+	_AST_GetWithReferences(clause, with_referred);
+	
+	_prepareIterateAll(with_aliases, &it);
+
+	// Check if every alias identifier that is referred was previously defined
+	while(raxNext(&it)) {
+		bool referred = false;
+		bool defined = false;
+		int len = it.key_len;
+		unsigned char *alias = it.key;
+		if(raxFind(with_referred, alias, len) != raxNotFound) {
+			referred = true;
+		}
+		if(raxFind(defined_aliases, alias, len) != raxNotFound) {
+			defined = true;
+		}
+		if(referred && !defined) {
+			ErrorCtx_SetError("%.*s not defined", len, alias);
+			res = AST_INVALID;
+			break;
+		}
+	}
```

Comment:
Also, I think we don't traverse all the referred aliases in this manner.. as they are in the expression of the projection of the WITH clause, which is not introduced to `defined_aliases` in `_AST_GetWithAliases()`.

Reply:
I think that iterate over the aliases collected by ```_AST_GetWithReferences()```was the first approach, but the program failed in cases like this:
https://github.com/RedisGraph/RedisGraph/blob/0354c688027a25df95128003dda4397afa800c24/tests/tck/features/expressions/list/List12.feature#L108

```
127.0.0.1:6379> GRAPH.QUERY g "CREATE (a:A), (a)-[:T]->(:B), (a)-[:T]->(:C)"
127.0.0.1:6379> GRAPH.QUERY g "MATCH p = (n:A)-->() WITH [x IN collect(p) | head(nodes(x))] AS p, count(n) AS c RETURN p, c"
```

defined_aliases [n, p]
with_aliases    [c, p]
with_referred   [n, p, x]

Although the query is correct, the validation will report an error because ```x``` is referred but it is neither an alias defined in the WITH clause nor an identifier defined previously.
If the function iterates over the identifiers collected in ```_AST_GetWithAliases()```: ```[c, p]```the validation is done correctly.



> Also, I think we don't traverse all the referred aliases in this manner..

I think that we don't need to traverse all the referred aliases, as noticed in the previous case. But if you think in some case where the function could fail, please tell me and I'll test it.
Thanks

Reply:
> The one thing that I have yet to understand, is why you skip the pattern comprehensions in line 1136?

Thanks.  It was an error.

There was not problem with the query:  ```GRAPH.QUERY g "WITH [x in li | x.prop1] AS li RETURN li"```
because it was not identified as pattern comprehension

But with the following query, the validation was not able to find that ```a``` was not defined:
```WITH [(n)-[x:R]->(m) | a.prop1] AS a RETURN 1```

Both queries where added to the tests.

Reply:
Ohh yes I switched list and path.. anyways you got the idea, great 👍